File: open_source_projects/Deepbox-main/benchmarks/python/12_stats.py
"""
Benchmark 12 — Statistics
SciPy / NumPy
"""

import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

import numpy as np
from scipy import stats as sp
from utils import run, create_suite, header, footer

suite = create_suite("stats", "SciPy")
header("Benchmark 12 — Statistics", "SciPy")

rng = np.random.RandomState(42)

a500 = rng.randn(500) * 10
b500 = rng.randn(500) * 10
a1k = rng.randn(1000) * 10
b1k = rng.randn(1000) * 10
a5k = rng.randn(5000) * 10
b5k = rng.randn(5000) * 10
a10k = rng.randn(10000) * 10
pos500 = np.abs(rng.randn(500)) * 10 + 0.1
pos1k = np.abs(rng.randn(1000)) * 10 + 0.1

# ── Descriptive Statistics ──────────────────────────────

run(suite, "mean", "500", lambda: np.mean(a500))
run(suite, "mean", "1K", lambda: np.mean(a1k))
run(suite, "mean", "10K", lambda: np.mean(a10k))
run(suite, "median", "500", lambda: np.median(a500))
run(suite, "median", "1K", lambda: np.median(a1k))
run(suite, "median", "10K", lambda: np.median(a10k))
run(suite, "mode", "500", lambda: sp.mode(a500, keepdims=False))
run(suite, "mode", "1K", lambda: sp.mode(a1k, keepdims=False))
run(suite, "std", "500", lambda: np.std(a500, ddof=1))
run(suite, "std", "1K", lambda: np.std(a1k, ddof=1))
run(suite, "std", "10K", lambda: np.std(a10k, ddof=1))
run(suite, "variance", "500", lambda: np.var(a500, ddof=1))
run(suite, "variance", "1K", lambda: np.var(a1k, ddof=1))
run(suite, "skewness", "500", lambda: sp.skew(a500))
run(suite, "skewness", "1K", lambda: sp.skew(a1k))
run(suite, "kurtosis", "500", lambda: sp.kurtosis(a500))
run(suite, "kurtosis", "1K", lambda: sp.kurtosis(a1k))
run(suite, "quantile (0.25)", "1K", lambda: np.quantile(a1k, 0.25))
run(suite, "quantile (0.75)", "1K", lambda: np.quantile(a1k, 0.75))
run(suite, "percentile (90)", "1K", lambda: np.percentile(a1k, 90))
run(suite, "geometricMean", "500", lambda: sp.gmean(pos500))
run(suite, "geometricMean", "1K", lambda: sp.gmean(pos1k))
run(suite, "harmonicMean", "500", lambda: sp.hmean(pos500))
run(suite, "harmonicMean", "1K", lambda: sp.hmean(pos1k))
run(suite, "trimMean (10%)", "1K", lambda: sp.trim_mean(a1k, 0.1))
run(suite, "moment (3rd)", "1K", lambda: sp.moment(a1k, 3))

# ── Correlation ─────────────────────────────────────────

run(suite, "pearsonr", "500", lambda: sp.pearsonr(a500, b500))
run(suite, "pearsonr", "1K", lambda: sp.pearsonr(a1k, b1k))
run(suite, "pearsonr", "5K", lambda: sp.pearsonr(a5k, b5k))
run(suite, "spearmanr", "500", lambda: sp.spearmanr(a500, b500))
run(suite, "spearmanr", "1K", lambda: sp.spearmanr(a1k, b1k))
run(suite, "kendalltau", "500", lambda: sp.kendalltau(a500, b500))
run(suite, "corrcoef", "500", lambda: np.corrcoef(a500, b500))
run(suite, "corrcoef", "1K", lambda: np.corrcoef(a1k, b1k))
run(suite, "cov", "500", lambda: np.cov(a500, b500))
run(suite, "cov", "1K", lambda: np.cov(a1k, b1k))

# ── Hypothesis Tests ────────────────────────────────────

run(suite, "ttest_1samp", "500", lambda: sp.ttest_1samp(a500, 0))
run(suite, "ttest_1samp", "1K", lambda: sp.ttest_1samp(a1k, 0))
run(suite, "ttest_ind", "500", lambda: sp.ttest_ind(a500, b500))
run(suite, "ttest_ind", "1K", lambda: sp.ttest_ind(a1k, b1k))
run(suite, "ttest_rel", "500", lambda: sp.ttest_rel(a500, b500))
run(suite, "ttest_rel", "1K", lambda: sp.ttest_rel(a1k, b1k))

rng2 = np.random.RandomState(1)
g1 = rng2.randn(200) * 10
rng3 = np.random.RandomState(2)
g2 = rng3.randn(200) * 10
rng4 = np.random.RandomState(3)
g3 = rng4.randn(200) * 10

run(suite, "f_oneway", "3×200", lambda: sp.f_oneway(g1, g2, g3))
run(suite, "chisquare", "10 bins", lambda: sp.chisquare([16, 18, 16, 14, 12, 12, 9, 10, 11, 12]))
run(suite, "shapiro", "500", lambda: sp.shapiro(a500))
run(suite, "mannwhitneyu", "500", lambda: sp.mannwhitneyu(a500, b500))
run(suite, "mannwhitneyu", "1K", lambda: sp.mannwhitneyu(a1k, b1k))
run(suite, "kruskal", "3×200", lambda: sp.kruskal(g1, g2, g3))
run(suite, "friedmanchisquare", "3×200", lambda: sp.friedmanchisquare(g1, g2, g3))
run(suite, "anderson", "500", lambda: sp.anderson(a500))
run(suite, "kstest", "500", lambda: sp.kstest(a500, "norm"))
run(suite, "kstest", "1K", lambda: sp.kstest(a1k, "norm"))
run(suite, "levene", "500", lambda: sp.levene(a500, b500))
run(suite, "bartlett", "500", lambda: sp.bartlett(a500, b500))
run(suite, "normaltest", "500", lambda: sp.normaltest(a500))
run(suite, "normaltest", "1K", lambda: sp.normaltest(a1k))
run(suite, "wilcoxon", "500", lambda: sp.wilcoxon(a500))

footer(suite, "scipy-stats.json")

File: open_source_projects/Deepbox-main/benchmarks/deepbox/12-stats.ts
/**
 * Benchmark 12 — Statistics
 * Deepbox vs SciPy / NumPy
 */

import { tensor } from "deepbox/ndarray";
import {
	anderson,
	bartlett,
	chisquare,
	corrcoef,
	cov,
	f_oneway,
	friedmanchisquare,
	geometricMean,
	harmonicMean,
	kendalltau,
	kruskal,
	kstest,
	kurtosis,
	levene,
	mannwhitneyu,
	mean,
	median,
	mode,
	moment,
	normaltest,
	pearsonr,
	percentile,
	quantile,
	shapiro,
	skewness,
	spearmanr,
	std,
	trimMean,
	ttest_1samp,
	ttest_ind,
	ttest_rel,
	variance,
	wilcoxon,
} from "deepbox/stats";
import { createSuite, footer, header, run } from "../utils";

const suite = createSuite("stats");
header("Benchmark 12 — Statistics");

// ── Data ────────────────────────────────────────────────

function seededRng(seed: number) {
	let s = seed >>> 0;
	return () => {
		s = (s * 1664525 + 1013904223) >>> 0;
		return s / 2 ** 32;
	};
}

function makeArray(n: number, seed: number): number[] {
	const rand = seededRng(seed);
	return Array.from({ length: n }, () => rand() * 20 - 10);
}

function makePositiveArray(n: number, seed: number): number[] {
	const rand = seededRng(seed);
	return Array.from({ length: n }, () => rand() * 10 + 0.1);
}

const a500 = tensor(makeArray(500, 42));
const b500 = tensor(makeArray(500, 99));
const a1k = tensor(makeArray(1000, 42));
const b1k = tensor(makeArray(1000, 99));
const a5k = tensor(makeArray(5000, 42));
const b5k = tensor(makeArray(5000, 99));
const a10k = tensor(makeArray(10000, 42));
const pos500 = tensor(makePositiveArray(500, 42));
const pos1k = tensor(makePositiveArray(1000, 42));

// ── Descriptive Statistics ──────────────────────────────

run(suite, "mean", "500", () => mean(a500));
run(suite, "mean", "1K", () => mean(a1k));
run(suite, "mean", "10K", () => mean(a10k));
run(suite, "median", "500", () => median(a500));
run(suite, "median", "1K", () => median(a1k));
run(suite, "median", "10K", () => median(a10k));
run(suite, "mode", "500", () => mode(a500));
run(suite, "mode", "1K", () => mode(a1k));
run(suite, "std", "500", () => std(a500));
run(suite, "std", "1K", () => std(a1k));
run(suite, "std", "10K", () => std(a10k));
run(suite, "variance", "500", () => variance(a500));
run(suite, "variance", "1K", () => variance(a1k));
run(suite, "skewness", "500", () => skewness(a500));
run(suite, "skewness", "1K", () => skewness(a1k));
run(suite, "kurtosis", "500", () => kurtosis(a500));
run(suite, "kurtosis", "1K", () => kurtosis(a1k));
run(suite, "quantile (0.25)", "1K", () => quantile(a1k, 0.25));
run(suite, "quantile (0.75)", "1K", () => quantile(a1k, 0.75));
run(suite, "percentile (90)", "1K", () => percentile(a1k, 90));
run(suite, "geometricMean", "500", () => geometricMean(pos500));
run(suite, "geometricMean", "1K", () => geometricMean(pos1k));
run(suite, "harmonicMean", "500", () => harmonicMean(pos500));
run(suite, "harmonicMean", "1K", () => harmonicMean(pos1k));
run(suite, "trimMean (10%)", "1K", () => trimMean(a1k, 0.1));
run(suite, "moment (3rd)", "1K", () => moment(a1k, 3));

// ── Correlation ─────────────────────────────────────────

run(suite, "pearsonr", "500", () => pearsonr(a500, b500));
run(suite, "pearsonr", "1K", () => pearsonr(a1k, b1k));
run(suite, "pearsonr", "5K", () => pearsonr(a5k, b5k));
run(suite, "spearmanr", "500", () => spearmanr(a500, b500));
run(suite, "spearmanr", "1K", () => spearmanr(a1k, b1k));
run(suite, "kendalltau", "500", () => kendalltau(a500, b500));
run(suite, "corrcoef", "500", () => corrcoef(a500, b500));
run(suite, "corrcoef", "1K", () => corrcoef(a1k, b1k));
run(suite, "cov", "500", () => cov(a500, b500));
run(suite, "cov", "1K", () => cov(a1k, b1k));

// ── Hypothesis Tests ────────────────────────────────────

run(suite, "ttest_1samp", "500", () => ttest_1samp(a500, 0));
run(suite, "ttest_1samp", "1K", () => ttest_1samp(a1k, 0));
run(suite, "ttest_ind", "500", () => ttest_ind(a500, b500));
run(suite, "ttest_ind", "1K", () => ttest_ind(a1k, b1k));
run(suite, "ttest_rel", "500", () => ttest_rel(a500, b500));
run(suite, "ttest_rel", "1K", () => ttest_rel(a1k, b1k));

const g1 = tensor(makeArray(200, 1));
const g2 = tensor(makeArray(200, 2));
const g3 = tensor(makeArray(200, 3));

run(suite, "f_oneway", "3×200", () => f_oneway(g1, g2, g3));
run(suite, "chisquare", "10 bins", () =>
	chisquare(tensor([16, 18, 16, 14, 12, 12, 9, 10, 11, 12]))
);
run(suite, "shapiro", "500", () => shapiro(a500));
run(suite, "mannwhitneyu", "500", () => mannwhitneyu(a500, b500));
run(suite, "mannwhitneyu", "1K", () => mannwhitneyu(a1k, b1k));
run(suite, "kruskal", "3×200", () => kruskal(g1, g2, g3));
run(suite, "friedmanchisquare", "3×200", () => friedmanchisquare(g1, g2, g3));
run(suite, "anderson", "500", () => anderson(a500));
run(suite, "kstest", "500", () => kstest(a500, "norm"));
run(suite, "kstest", "1K", () => kstest(a1k, "norm"));
run(suite, "levene", "500", () => levene("median", a500, b500));
run(suite, "bartlett", "500", () => bartlett(a500, b500));
run(suite, "normaltest", "500", () => normaltest(a500));
run(suite, "normaltest", "1K", () => normaltest(a1k));
run(suite, "wilcoxon", "500", () => wilcoxon(a500));

footer(suite, "deepbox-stats.json");
